jwkscache/cache.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	return nil
+}
+
+func (c *JWKSCache) initJWKSFromURL(ctx context.Context, url string) error {
+	// Create the JWKS cache
+	cache := jwk.NewCache(ctx,
+		jwk.WithErrSink(httprc.ErrSinkFunc(func(err error) {
+			c.logger.Warnf("Error while refreshing JWKS cache: %v", err)
+		})),
+	)
+
+	// We also need to create a custom HTTP client (if we don't have one already) because otherwise there's no timeout.
+	if c.client == nil {
+		c.client = &http.Client{
+			Timeout: c.requestTimeout,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					MinVersion: tls.VersionTLS12,
+				},
+			},
+		}
+	}
+
+	// Register the cache
+	err := cache.Register(url,
+		jwk.WithMinRefreshInterval(c.minRefreshInterval),
+		jwk.WithHTTPClient(c.client),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to register JWKS cache: %w", err)
+	}
+
+	// Fetch the JWKS right away to start, so we can check it's valid and populate the cache
+	refreshCtx, refreshCancel := context.WithTimeout(ctx, c.requestTimeout)
+	_, err = cache.Refresh(refreshCtx, url)
+	refreshCancel()
+	if err != nil {
+		return fmt.Errorf("failed to fetch JWKS: %w", err)
+	}
+
+	c.jwks = jwk.NewCachedSet(cache, url)
+	return nil
+}
+
+func (c *JWKSCache) initJWKSFromFile(ctx context.Context, file string) error {
+	// Get the path to the folder containing the file
+	path := filepath.Dir(file)
+
+	// Start watching for changes in the filesystem
+	eventCh := make(chan struct{})
+	loaded := make(chan error, 1) // Needs to be buffered to prevent a goroutine leak
+	go func() {
+		watchErr := fswatcher.Watch(ctx, path, eventCh)
+		if watchErr != nil && !errors.Is(watchErr, context.Canceled) {
+			// Log errors only
+			c.logger.Errorf("Error while watching for changes to the local JWKS file: %v", watchErr)
+		}
+	}()
+	go func() {
+		var firstDone bool
+		for {
+			select {
+			case <-eventCh:
+				// When there's a change, reload the JWKS file
+				err := c.parseJWKSFile(file)
+				if !firstDone {
+					// The first time, signal that the initialization was complete and pass the error
+					loaded <- err
+					close(loaded)
+					firstDone = true
+				} else if err != nil {
+					// Log errors only
+					c.logger.Errorf("Error reading JWKS from disk: %v", err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Trigger a refresh immediately and wait for the first reload
+	eventCh <- struct{}{}
+
+	select {
+	case err := <-loaded:
+		// Error could be nil if everything is fine
+		return err
+	case <-time.After(5 * time.Second):
+		// If we don't get a response in 5s, something bad's going on
+		return errors.New("failed to initialize JWKS from file: no file loaded after 5s")
+	case <-ctx.Done():
+		return fmt.Errorf("failed to initialize JWKS from file: %w", ctx.Err())
+	}
+}
+
+// Used by initJWKSFromFile to parse a JWKS file every time it's changed
+func (c *JWKSCache) parseJWKSFile(file string) error {
+	c.logger.Debugf("Reloading JWKS file from disk")
```

Comment:
The first time is not "reloading", "parsing" is better.

Reply:
Done

Reply:
(I have moved it a few lines above where we have `firstDone`)